Sources/Algorithms/Keyed.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct KeysAreNotUnique<Key, Element>: Error {
+  public let key: Key
+  public let previousElement: Element
+  public let conflictedElement: Element
+
+  public init(key: Key, previousElement: Element, conflictedElement: Element) {
+    self.key = key
+    self.previousElement = previousElement
+    self.conflictedElement = conflictedElement
+  }
+}
+
+extension Sequence {
+  /// Creates a new Dictionary from the elements of `self`, keyed by the
+  /// results returned by the given `keyForValue` closure. Deriving the
```

Comment:
It might help to say which specific error type is thrown.  Plus this can be moved to the formal `throws` DocC field, a la:

```swift
/// - Throws: `KeysAreNotUnique ` if two values map to the same key (via `keyForValue`).
```

Reply:
Interesting, there are no other thrown errors in the standard library, or even `swift-algorithms`. Should this be a documented error that we expect users to catch, or should it be treated more like a softer version of a `fatalError`?

Reply:
I think it should be documented, mainly because callers need to be able to reliably distinguish between errors thrown by this library method itself vs deeper errors from the closures, because they might want to handle those cases differently (e.g. distinguish to the user that the problem is duplicate keys vs an invalid value, or failure to fetch some additional data from the network, etc).

As to this being novel for this library, that might just be coincidence?  Perhaps swift-algorithms just doesn't happen to contain a lot of cases where there is an error condition (that can't otherwise be handled implicitly, such as returning nil).

Looking at [all the uses of `fatalError` in this library](https://github.com/search?q=repo%3Aapple%2Fswift-algorithms%20fatalError&type=code), they're all related to indexing errors.  Those are a different breed, I think - they're more likely to represent actual code bugs, for a start.

Pragmatically, perhaps the key decision point is how much work it is to pre-emptively validate the data vs just catching an exception.  For subscripting, for example, it's arguably easy & fast enough to just check if your index is within range first.  A similar argument can be made for why the basic arithmetic operators in Swift will just crash on overflow / underflow / etc rather than throw exceptions.  But for more complex operations - like `group(by:)` and `keyed(by:)`, pre-emptively checking for duplicate keys means basically reimplementing those methods.  I don't think that's reasonable to expect of callers.

#### Sidebar on crashing philosophy

There's a quiet philosophical debate within the Swift community about how to handle errors.  Some folks think it's preferable to just crash (and thus save API users from having to deal with exceptions), while others (myself included) prefer to give callers control over whether they crash or handle errors in some other manner.

Since crashing can be achieved with just a suitably-placed `try!`, there's very little added cost for people that want their application to crash instead of having to handle the exception.

For most use-cases, crashing just isn't acceptable (e.g. many server applications, most end-user applications, etc).  Bad input data should just fail the relevant, specific operation, not the entire application.  e.g. even if it's the Save operation that's broken by the exception, letting the application keep running at least gives the user a chance to save their data in other ways (e.g. copy-pasting it into another application, or dragging it out to a file in the Finder, etc).

More broadly, crashing is almost always an unsafe operation - who knows what things in the system, like databases or other data stores on disk, are in an invalid state right at that moment.  The stack needs to be gracefully unwound in order to ensure persistent state is not corrupted.